src/mds/ScrubStack.cc

```diff
@@ -0,0 +1,435 @@
+    scrub_dirfrag(cur_dir, &frag_added_children, &frag_terminal, &frag_done);
+    if (frag_done) {
+      // FIXME is this right?  Can we end up hitting this more than
+      // once and is that a problem?
+      cur_dir->inode->scrub_dirfrag_finished(cur_dir->frag);
+    }
+    *added_children |= frag_added_children;
+    all_frags_terminal = all_frags_terminal && frag_terminal;
+    all_frags_done = all_frags_done && frag_done;
+  }
+
+  dout(20) << "finished looping; all_frags_terminal=" << all_frags_terminal
+           << ", all_frags_done=" << all_frags_done << dendl;
+  if (all_frags_done) {
+    assert (!*added_children); // can't do this if children are still pending
+
+    // OK, so now I can... fire off a validate on the dir inode, and
+    // when it completes, come through here again, noticing that we've
+    // set a flag to indicate the the validate happened, and 
+
+    scrub_dir_dentry_final(dn, &finally_done);
+  }
+
+  *terminal = all_frags_terminal;
+  *done = all_frags_done && finally_done;
+  dout(10) << __func__ << " is exiting " << *terminal << " " << *done << dendl;
+  return;
+}
+
+bool ScrubStack::get_next_cdir(CInode *in, CDir **new_dir)
+{
+  dout(20) << __func__ << " on " << *in << dendl;
+  frag_t next_frag;
+  int r = in->scrub_dirfrag_next(&next_frag);
+  assert (r >= 0);
+
+  if (r == 0) {
+    // we got a frag to scrub, otherwise it would be ENOENT
+    dout(25) << "looking up new frag " << next_frag << dendl;
+    CDir *next_dir = in->get_or_open_dirfrag(mdcache, next_frag);
+    if (!next_dir->is_complete()) {
+      C_KickOffScrubs *c = new C_KickOffScrubs(mdcache->mds, this);
+      next_dir->fetch(c);
+      dout(25) << "fetching frag from RADOS" << dendl;
+      return false;
+    }
+    *new_dir = next_dir;
+    dout(25) << "returning dir " << *new_dir << dendl;
+    return true;
+  }
+  assert(r == ENOENT);
+  // there are no dirfrags left
+  *new_dir = NULL;
+  return true;
+}
+
+class C_InodeValidated : public MDSInternalContext
+{
+  public:
+    ScrubStack *stack;
+    CInode::validated_data result;
+    CDentry *target;
+
+    C_InodeValidated(MDSRank *mds, ScrubStack *stack_, CDentry *target_)
+      : MDSInternalContext(mds), stack(stack_), target(target_)
+    {}
+
+    void finish(int r)
+    {
+      stack->_validate_inode_done(target, r, result);
+    }
+};
+
+
+void ScrubStack::scrub_dir_dentry_final(CDentry *dn, bool *finally_done)
+{
+  dout(20) << __func__ << *dn << dendl;
+
+  // Two passes through this function.  First one triggers inode validation,
+  // second one sets finally_done
+  // FIXME: kind of overloading scrub_in_progress here, using it while
+  // dentry is still on stack to indicate that we have finished
+  // doing our validate_disk_state on the inode
+  // FIXME: the magic-constructing scrub_info() is going to leave
+  // an unneeded scrub_infop lying around here
+  *finally_done = !dn->scrub_info()->dentry_scrubbing;
+  if (!*finally_done) {
```

Comment:
dentry_scrubbing is set to true on initiating. (no space updates its value). this means scrub_dir_dentry_final() can call  in->validate_disk_state() multiple times